Merge reduced LDMX tracking producer

fdelzanno opened the pull request:
I am updating _ldmx-sw_, here are the details.

### What are the issues that this addresses?
<!--
_Hint_: Use the phrase '_This resolves #< issue number >_' so that they are linked automatically.
-->

This resolves iss1491-a-reduced-tracking by adding 3 files to the tracking branch of ldmx-sw: a track seeder, a track finder, and a truth tracker optimized for the reduced LDMX geometry.

## Check List
- [x] I successfully compiled _ldmx-sw_ with my developments
- [x] I ran my developments and the following shows that they are successful.
See comments below for summary plots and descriptions of files.

fdelzanno committed:
Test Commit

fdelzanno committed:
This is the first update to the reduced tracking code (4:01 pm, 11/08/24). I added reducedSeeding, created the reducedTrack class, and modified the tracking.py to allow reducedSeeding to run.

fdelzanno committed:
Second update to reduced tracking code (6:27 pm, 11/11/24). ReducedSeeding now runs as a producer in ldmx-sw and saves data to a ReducedTrack object.

fdelzanno committed:
ReducedSeeding now runs as a producer in ldmx-sw and saves data to a ReducedTrack object.

fdelzanno committed:
Formatting

fdelzanno committed:
Merge branch 'iss1491-a-reduced-tracking' of github.com:LDMX-Software/ldmx-sw into iss1491-a-reduced-tracking

fdelzanno committed:
Case Renaming for Certain Files

fdelzanno committed:
ReducedTrackFinder implemented and running

fdelzanno committed:
Adding an event display python script to the Tracking/python

fdelzanno committed:
Delete event_display.py

Wrong event display...

fdelzanno committed:
Adding the correct eventDisplay code to Tracking/python

fdelzanno committed:
Merge branch 'iss1491-a-reduced-tracking' of github.com:LDMX-Software/ldmx-sw into iss1491-a-reduced-tracking

fdelzanno committed:
Updated ReducedTrackFinder logic: We select the best seed, then remove any seed that has ANY overlapping digiRecoilHit with the best seed. Originally we were only removing those that had overlapping averaged hits. Need to implement a onEndProcess statement that counts all of the tracks (currently, only outputs for each track itearction).

fdelzanno committed:
Restructured main tracking codes to incorporate the 'ReducedTrack' object instead of dealing with predefined arrays

fdelzanno committed:
Fixed return statement in ReducedTrackFinder.cxx to correctly return 'bestTracks' vector

fdelzanno committed:
Code restructure to permit the inclusion of the trackID calculation. This truth level information is calculated at the Seeder level, and then is associated to a seed. The track finding does not use any trackID information when selecting the track, but the trackID of the chosen track is available for truth-level matching.

fdelzanno committed:
Two updates: 1) renamed tracking files to remove mention of 'reduced', to avoid confusing terminology. The new track object is 'StraightTrack' and the processors are 'Linear'; 2) added a 'TruthTrack' processor that fits only hits with truthID == 1. This should allow straightforward comparison between tracking and truth info.

fdelzanno committed:
Fixed nullptr issue in TruthTracking, added TruthTracks to event display code

fdelzanno committed:
TruthTrack no longer dependant on existence of recHits.

fdelzanno committed:
Minor bug fix: require the truth track to be pure (truthProb = 1 for the 4-hit track)

fdelzanno committed:
Cosmetic Changes. Keep track of sensor locations in .py files to facilitate changing geometry. Added TS simulation to example config. Small updates to eventDisplay code.

fdelzanno committed:
Added a covariance matrix to the fitting algorithm so we can have uncertainties on fitting parameters. Also added a dqm sequence explicitly for straight tracks that can be added to an LDMX sim. Miscellaneous cleanup to make code more readable.

fdelzanno committed:
Fixed issue: when we don't have enough recoil sim hits, we now output an empty track to the collection instead of skipping the event. This way the collection is created in cases where the first event of the simulation doesn't have enough sim info for track reconstruction

fdelzanno committed:
Fix handling of cases in which we have less than 4 unique tracker hits. Fit if layer A and layer B both have a hit, use the weighted average sensor point in the fit. Otherwise, take whichever of the two layers has a hit to be the 'effective' sensor point.

fdelzanno committed:
Add rLDMX tracking into CI script sequence

fdelzanno commented:
I ran a 5,000 event sample using a 4GeV eGun in the reduced-ldmx-v1 geometry to verify that the tracking is working. Here are the statistics:

There are 5000 events in the sample.
We find 5008 tracks / 4703 events with tracks + 297 empty events in which we don't find a track = **5008 tracks / 5000 events**.
We find 4808 truth tracks / 4808 events with truth tracks + 192 empty events in which we don't fit a truth track = **4808 tracks / 5000 events**

Out of the tracks in the sample, **4760 tracks** have trackID = 1 + **248 tracks** have a different trackID number = 5008 tracks

Out of the tracks with trackID = 1 in the sample:
- there are 4567 that have truthProb = 1 (i.e. all 4 hits used in tracking correspond to the recoil electron)
- 129 tracks that have truthProb = 0.75 (i.e. 3 hits / 4 used in tracking correspond to recoil electron)
- 64 tracks that have truthProb = 0.5 (i.e. 2 hits / 4 used in tracking correspond to recoil electron)

We reconstruct 4895 electron tracks + 112 positron tracks + 1 pion track = 5008 tracks/5008.

Here are the summary plots. In blue are all the reconstructed tracks. In green are the truth tracks. In red is the data only from reconstructed tracks with trackID = 1 (i.e. reconstructed recoil electron tracks).

![Image](https://github.com/user-attachments/assets/9fb379f6-0ed7-442c-ac03-ce6abd9251f3)

![Image](https://github.com/user-attachments/assets/826b42ac-e233-4885-86d9-dc808a4eb767)

It makes sense that the chi2 is lower for reconstructed tracks, since the tracking looks for tracks with the lowest chi2.

![Image](https://github.com/user-attachments/assets/83dab5f0-63e8-40eb-9444-11822913777e)

This is the distance between the track extrapolated to the first Ecal layer and the RecHit used to create the track. The threshold value is 15 mm, so there is no reconstructed track with dRecHit > 15 mm. The radius of an Ecal cell is approximately 4 mm, meaning any track with dRecHit < 4 mm is likely to be in the correct cell. 

I also have the pulls of the reconstructed track beam (theta) and azimuthal (phi) angle (these come from the dqm, and are for unique tracks):

<img width="1038" alt="Image" src="https://github.com/user-attachments/assets/7d6a0c9a-dc4e-412c-85db-4b4c6850dc39" />

<img width="1014" alt="Image" src="https://github.com/user-attachments/assets/a6bcca88-328c-4d87-83ce-d22049cf1ec2" />

For events in which we construct a truth track, I plotted the difference between the location of the trackID = 1 reconstructed track and the truth track at both the Target and Ecal face. 

![Image](https://github.com/user-attachments/assets/8aa2b264-9384-456a-93c6-da60b84a801b)

![Image](https://github.com/user-attachments/assets/f82f072a-6a2d-440e-ac1f-1f70e07ad06d)

We expect a greater spread in y since it is the more uncertain direction.

![Image](https://github.com/user-attachments/assets/fb558d77-9d93-4cc8-a8ca-76aed0153dc0)

![Image](https://github.com/user-attachments/assets/45a175a6-d02b-4761-8654-f51d4d54eef1)

fdelzanno committed:
Merge branch 'trunk' into iss1491-a-reduced-tracking

github-actions[bot] committed:
Apply clang-format

tvami commented on .github/validation_samples/reduced_ldmx/config.py:
```diff
@@ -68,6 +68,46 @@
+digiRecoil.hit_collection = "RecoilSimHits"
+digiRecoil.out_collection = "DigiRecoilSimHits"
+digiRecoil.merge_hits = True
+digiRecoil.sigma_u = uSmearing
+digiRecoil.sigma_v = vSmearing
```
```suggestion
digiRecoil.sigma_u = 0.006
digiRecoil.sigma_v = 0.000001
```

tvami commented on .github/validation_samples/reduced_ldmx/config.py:
```diff
@@ -68,6 +68,46 @@
+from LDMX.Tracking.geo import TrackersTrackingGeometryProvider as trackgeo
+trackgeo.get_instance().setDetector(det)
+
+#smearings
+uSmearing = 0.006       #mm #could bump up to 10 micron if we want
+vSmearing = 0.000001    #mm #~unused
+
```
```suggestion
```

tvami commented on Detectors/test/reducedLDMX_eGun_config.py:
```diff
@@ -26,11 +26,11 @@
 p.termLogLevel = 0
 
```
This file is somewhat historical, and was made before we made it into a CI. I'd not add the tracking stuff in here (and actually would delete this file to avoid future confusion)

fdelzanno replied:
I agree. I removed the file from the branch.

tvami replied:
I dont think this went thru, did you do `git rm Detectors/test/reducedLDMX_eGun_config.py` ?

tvami commented on Tracking/exampleConfigs/test_reduced_config.py:
```diff
@@ -0,0 +1,123 @@
+digiRecoil.sigma_v = vSmearing
+
+#Assuming rLDMX-v1 geometry. These should be updated if one changes the position of the recoil layers
+#TODO: is there some way to get this directly from the gdml?
```
There isnt :(

fdelzanno replied:
How sad!

tvami commented on Tracking/exampleConfigs/test_reduced_config.py:
```diff
@@ -0,0 +1,123 @@
+layer23_mid = (15.5+24.5)/2.
+layer34_mid = (24.5+30.5)/2.
+
+truth_tracking = tracking.LinearTruthTracking("LinearTruthTracking")
+truth_tracking.input_hit_collection = "DigiRecoilSimHits"
+truth_tracking.input_recHits_collection = "EcalRecHits"
+truth_tracking.out_track_collection = "LinearRecoilTruthTracks"
+truth_tracking.layer12_midpoint = layer12_mid
+truth_tracking.layer23_midpoint = layer23_mid
+truth_tracking.layer34_midpoint = layer34_mid
```
Why is this part not in the CI script? 
Also I have similar comments of having this info in the CI and not having yet another config

tvami commented on Tracking/include/Tracking/Event/Measurement.h:
```diff
@@ -92,6 +92,8 @@ class Measurement {
   /// @return The hit time in ns.
   [[nodiscard]] float getTime() const { return t_; };
 
+  [[nodiscard]] float getEdep() const { return edep_; };
+
```
where is this used in the new code?

fdelzanno replied:
Good catch. I'm not sure where this came from, but I removed it because it shouldn't be there.

tvami commented on Tracking/include/Tracking/Event/StraightTrack.h:
```diff
@@ -0,0 +1,187 @@
+  std::vector<double> trk_cov_;
+
+  /// Class declaration needed by the ROOT dictionary.
+  ClassDef(StraightTrack, 3);
```
```suggestion
  ClassDef(StraightTrack, 1);
```

tvami commented on Tracking/include/Tracking/Event/StraightTrack.h:
```diff
@@ -0,0 +1,187 @@
+  void setCov(const std::vector<double>& cov) { trk_cov_ = cov; }
+  std::vector<double> getCov() const { return trk_cov_; }
+
+ protected:
```
please rename the variables to `snake_case`, or for the member variables `snake_case_`.
Generally here are the style rules: https://ldmx-software.github.io/developing/coding-rules.html

fdelzanno replied:
I went through and renamed all the variables in each file (except the eventDisplay.py code) according to the style rules. Hopefully I didn't miss anything...

tvami commented on Tracking/include/Tracking/Reco/LinearSeedFinder.h:
```diff
@@ -0,0 +1,144 @@
+  LinearSeedFinder(const std::string &name, framework::Process &process);
+
+  /// Destructor
+  ~LinearSeedFinder();
```
See III.3 https://ldmx-software.github.io/developing/coding-rules.html#iii-code-quality

fdelzanno replied:
I think I resolved this according to the coding rule. Please check again and let me know if I misinterpreted it.

tvami commented on Tracking/include/Tracking/Reco/LinearSeedFinder.h:
```diff
@@ -0,0 +1,144 @@
+  /// Destructor
+  ~LinearSeedFinder();
+
+  /**
+   *
+   */
+  void onProcessStart() override;
+
+  /**
+   *
+   */
+  void onProcessEnd() override;
```
```suggestion
```

tvami commented on Tracking/include/Tracking/Reco/LinearSeedFinder.h:
```diff
@@ -0,0 +1,144 @@
+ protected:
+  // Function to find seeds based on 2 Recoil points and 1 EcalRecHit
+  ldmx::StraightTrack SeedTracker(
+      const std::tuple<std::array<double, 3>, ldmx::Measurement,
```
can we make that "3" a variable? If in some lucky case we get to have another tracking layer?

fdelzanno replied:
The "3" in this case doesn't relate to the number of layers we have to fit. It just signifies that the array contains three position coordinates z, x, y.

I structured the function to build a recoil tuple that contains: an array<double, 3> which encodes the sensor position, an ldmx::Measurement that contains the actual ldmx reference to the first layer that was used, and an std::optional<ldmx::Measurement>, which is either the second layer that was used or a nullptr if there was no other layer with which to do the weighted averaging. The sensor position is passed directly to the fitting function along with the other recoil sensor position (two total) and the EcalRecHit position, while the two ldmx::Measurements are used to retrieve truth information from the Sim level.

If we were to have another tracking layer, all we would need to do is the add a third recoil tuple as a parameter and switch the third element in the fitting function from the Ecal position to the third recoil sensor position (as well as make some small adjustments to the uncertainties that are assigned and probably edit the StraightTracks class to save a third sensor position). I can add those references now but they would be unused in the current implementation.

tvami replied:
It's ok let's keep this as is

tvami commented on Tracking/include/Tracking/Reco/LinearSeedFinder.h:
```diff
@@ -0,0 +1,144 @@
+                         const std::array<double, 3> &ecalHit, double ax,
+                         double ay, double bx, double by);
+
+  // Function to find the number of unique layers hit (to determine if we have
+  // enough points to fit)
+  int uniqueLayersHit(const std::vector<ldmx::Measurement> &digiPoints);
```
I dont see this used anywhere?
```suggestion
```

fdelzanno replied:
The function is used in line 63 of the LinearSeedFinder.cxx. It's just a check to make sure we actually have enough distinct layers to look for a track.

tvami replied:
Ahh I see it, github did not expand that file in default

tvami commented on Tracking/include/Tracking/Reco/LinearSeedFinder.h:
```diff
@@ -0,0 +1,144 @@
+  std::string input_recHits_collection_{"EcalRecHits"};
+
+  double ecal_uncertainty_{3.87};
+  double ecal_distance_threshold_{
```
```suggestion
// Max distance from RecHit for valid track
  double ecal_distance_threshold_{10.0};  
```

tvami commented on Tracking/include/Tracking/Reco/LinearSeedFinder.h:
```diff
@@ -0,0 +1,144 @@
+
+  double ecal_uncertainty_{3.87};
+  double ecal_distance_threshold_{
+      10.0};  // Max distance from RecHit for valid track
```
```suggestion
```
if you just accepted the above suggestion this line needs to be removed

tvami commented on Tracking/python/eventDisplay.py:
```diff
@@ -0,0 +1,265 @@
+from LDMX.Framework import EventTree
```
I wonder if this would be better under https://github.com/LDMX-Software/ldmx-sw/tree/trunk/EventDisplay

fdelzanno replied:
I think it would make sense to put it in the EventDisplay, but in the README for the EventDisplay directory, it says that the directory was abandoned. I'm not sure it makes sense to store usable code in an abandoned directory (where people might not think to look for/use it)

tvami replied:
It's true that the c++ in there is abandoned, but you can edit the README to point out that this file is working. People would also not look for event display under the Tracking :D

fdelzanno replied:
Haha fair enough. I will move the file to event display and update the README.

tvami commented on Tracking/python/tracking.py:
```diff
@@ -44,6 +44,105 @@ def __init__(self, instance_name="DigitizationProcessor"):
         self.hit_collection = 'TaggerSimHits'
         self.out_collection = 'OutputMeasurements'
                 
+class LinearSeedFinder(Producer):
```
all these might be better in its own file `Tracking/python/reducedTracking.py`

fdelzanno replied:
I created a reducedTracking.py to fit these declarations, and updated the config to reflect the change.

tvami commented on Tracking/src/Tracking/Event/StraightTrack.cxx:
```diff
@@ -0,0 +1,11 @@
+
+    namespace ldmx {
+  void StraightTrack::Print() const {
+    std::cout << "print straight track" << std::endl;
```
Hmm but you are not printing it out (yet)

fdelzanno replied:
I added this file because I was following the logic of the track class, which has a Track.cxx in which the only line is to print the track (although I've never used it). If this file is unnecessary I can remove it from the branch (?).

tvami replied:
Well at one point it should be cleaned in the Track.cxx too, but let's just not add it first

fdelzanno replied:
Ok, I removed the file from the branch.

tvami commented on Tracking/src/Tracking/Reco/LinearTruthTracking.cxx:
```diff
@@ -0,0 +1,308 @@
+
+  if (ecalPoints.size() > 0) {
+    double ecal_firstLayer_z =
+        ecalPoints[0][0];  // Z position from the first point in ecalPoints
```
put the comment above here too

tvami reviewed (changes requested):
This is an amazing work @fdelzanno thanks! I have a few requests, nothing too big, but I didnt always point to it, so if the same thing is true in a different file, please change it there too

fdelzanno committed:
Update Tracking/include/Tracking/Reco/LinearSeedFinder.h

Co-authored-by: Tamas Vami <tamas.almos.vami@cern.ch>

fdelzanno committed:
Remove old rLDMX config, miscellaneous cleanup

fdelzanno committed:
Merge branch 'iss1491-a-reduced-tracking' of github.com:LDMX-Software/ldmx-sw into iss1491-a-reduced-tracking

github-actions[bot] committed:
Apply clang-format

fdelzanno committed:
Remove old rLDMX config, miscellaneous cleanup

fdelzanno committed:
Update .github/validation_samples/reduced_ldmx/config.py

Co-authored-by: Tamas Vami <tamas.almos.vami@cern.ch>

fdelzanno committed:
Update .github/validation_samples/reduced_ldmx/config.py

Co-authored-by: Tamas Vami <tamas.almos.vami@cern.ch>

fdelzanno committed:
Update Tracking/include/Tracking/Event/StraightTrack.h

Co-authored-by: Tamas Vami <tamas.almos.vami@cern.ch>

fdelzanno committed:
Update Tracking/include/Tracking/Reco/LinearSeedFinder.h

Co-authored-by: Tamas Vami <tamas.almos.vami@cern.ch>

fdelzanno committed:
Update Tracking/include/Tracking/Reco/LinearSeedFinder.h

Co-authored-by: Tamas Vami <tamas.almos.vami@cern.ch>

github-actions[bot] committed:
Apply clang-format

fdelzanno committed:
Commit to fix merge issue

fdelzanno committed:
Merge branch 'iss1491-a-reduced-tracking' of github.com:LDMX-Software/ldmx-sw into iss1491-a-reduced-tracking

github-actions[bot] committed:
Apply clang-format

fdelzanno committed:
Miscellaneous changes to address PR comments. Created a reducedTracking.py to hold all reduced Tracking producers. Modified all files to follow LDMX coding practice and style rules.

fdelzanno committed:
Merge branch 'iss1491-a-reduced-tracking' of github.com:LDMX-Software/ldmx-sw into iss1491-a-reduced-tracking

github-actions[bot] committed:
Apply clang-format

tvami commented on Tracking/src/Tracking/Reco/LinearTrackFinder.cxx:
```diff
@@ -0,0 +1,163 @@
+
+void LinearTrackFinder::onProcessStart() {}
+
+void LinearTrackFinder::configure(framework::config::Parameters& parameters) {
```
You should have the pass name in here too

tvami commented on Tracking/src/Tracking/Reco/LinearTrackFinder.cxx:
```diff
@@ -0,0 +1,163 @@
+                                     framework::Process& process)
+    : TrackingGeometryUser(name, process) {}
+
+void LinearTrackFinder::onProcessStart() {}
```
I dont think this is used 
```suggestion
```

fdelzanno replied:
You're right. I removed the function.

tvami commented on Tracking/src/Tracking/Reco/LinearTruthTracking.cxx:
```diff
@@ -0,0 +1,307 @@
+  truth_matching_tool_ = std::make_shared<tracking::sim::TruthMatchingTool>();
+}
+
+void LinearTruthTracking::configure(framework::config::Parameters& parameters) {
```
Pass name here too

tvami commented on Tracking/src/Tracking/Reco/LinearTruthTracking.cxx:
```diff
@@ -0,0 +1,307 @@
+  std::vector<std::array<double, 3>> first_layer_ecal_rec_hits;
+
+  for (const auto& x_ecal : ecal_rec_hit) {
+    if (x_ecal.getZPos() < 250) {
```
that 250 should be not hardcoded

fdelzanno replied:
Added a ecal_firstLayerZ_threshold variable and set it's value to 250, so that we can easily change the condition if we move the Ecal.

tvami replied:
that's good, except that the it should be `ecal_first_layer_z_threshold` according to the `snake_case` haha

fdelzanno replied:
True, fixed!

tvami commented on Tracking/src/Tracking/dqm/StraightTracksDQM.cxx:
```diff
@@ -0,0 +1,386 @@
+
+namespace tracking::dqm {
+
+void StraightTracksDQM::configure(framework::config::Parameters& parameters) {
```
pass name here too

tvami commented on Tracking/include/Tracking/Reco/LinearTrackFinder.h:
```diff
@@ -0,0 +1,84 @@
+
+  /// Destructor
+  virtual ~LinearTrackFinder() = default;
+  /**
+   *
+   */
+  void onProcessStart() override;
+  /**
+   *
+   */
+  void onProcessEnd() override;
```
```suggestion
```

fdelzanno committed:
Address PR comments, clean up code, move and delete files

fdelzanno committed:
Resolved Merge Conflicts

github-actions[bot] committed:
Apply clang-format

tvami commented on Tracking/src/Tracking/Reco/LinearSeedFinder.cxx:
```diff
@@ -0,0 +1,414 @@
+  }    // for positions in ecalRecHit
+
+  // Check if we would fit empty seeds, if so: end tracking
+  if (recoil_hits.size() < 2 || first_layer_ecal_rec_hits.empty() ||
+      uniqueLayersHit(recoil_hits) < 2) {
+    n_missing_++;
```
I ran into parenthesis issues in the past for sure this is safer
```suggestion
  if ((recoil_hits.size() < 2) || (first_layer_ecal_rec_hits.empty()) ||
      (uniqueLayersHit(recoil_hits) < 2)) {
    n_missing_++;
```

fdelzanno replied:
Fixed in Commit 675f482.

tvami commented on Tracking/src/Tracking/dqm/StraightTracksDQM.cxx:
```diff
@@ -0,0 +1,388 @@
+  input_pass_name_ =
+      parameters.getParameter<std::string>("input_pass_name", "");
+
+  ldmx_log(info) << "Track Collection " << track_collection_ << std::endl;
```
You dont need the `<< std::endl` with the `ldmx_log`. I think you removed it in every other file, except for this one. Please remove it here (in all 15 occurrences)

fdelzanno replied:
Fixed in Commit 675f482.

tvami commented on Tracking/src/Tracking/dqm/StraightTracksDQM.cxx:
```diff
@@ -0,0 +1,388 @@
+  // The total number of elements in the uniqueTracks and duplicateTracks
+  // vectors should be equal to the number of elements in the original tracks
+  // vector
+  if (unique_tracks.size() + duplicate_tracks.size() + fake_tracks.size() !=
```
```suggestion
  if ((unique_tracks.size() + duplicate_tracks.size() + fake_tracks.size()) !=
```

fdelzanno replied:
Fixed in Commit 675f482.

tvami commented on Tracking/src/Tracking/dqm/StraightTracksDQM.cxx:
```diff
@@ -0,0 +1,388 @@
+  // vector
+  if (unique_tracks.size() + duplicate_tracks.size() + fake_tracks.size() !=
+      tracks.size()) {
+    std::cerr << "Error: unique and duplicate tracks vectors do not add up to "
```
```suggestion
    ldmx_log(error) << "Unique and duplicate tracks vectors do not add up to "
```

fdelzanno replied:
Fixed in Commit 675f482.

tvami commented on Tracking/src/Tracking/dqm/StraightTracksDQM.cxx:
```diff
@@ -0,0 +1,388 @@
+    return;
+  }  // if different tracks don't add up to correct total
+
+  if (debug_) {
+    // Iterate through the uniqueTracks vector and duplicateTracks vector
+    std::cout << "Unique tracks:" << std::endl;
+    for (const ldmx::StraightTrack& track : unique_tracks) {
+      std::cout << "Track ID: " << track.getTrackID()
+                << ", Truth Prob: " << track.getTruthProb() << std::endl;
+    }
+    std::cout << "Duplicate tracks:" << std::endl;
+    for (const ldmx::StraightTrack& track : duplicate_tracks) {
+      std::cout << "Track ID: " << track.getTrackID()
+                << ", Truth Prob: " << track.getTruthProb() << std::endl;
+    }
+    std::cout << "Fake tracks:" << std::endl;
+    for (const ldmx::StraightTrack& track : fake_tracks) {
+      std::cout << "Track ID: " << track.getTrackID()
+                << ", Truth Prob: " << track.getTruthProb() << std::endl;
+    }
```
ok this is super new, I didnt even get to add it in the documentation, but we introduced `ldmx_log(trace)` for things that are more verbose than `ldmx_log(debug)`, so I suggest to remove the `debug_` boolean and move all this stuff to `ldmx_log(trace)` [and I'll update the documentation about this now]

fdelzanno replied:
Ok, basically I removed the debug_ boolean from both the .cxx and .h dqm, and turned these statements into ldmx_log(trace). Otherwise the style and content of the statements is identical.

Addressed in Commit 675f482.

tvami reviewed (changes requested):
I got to the DQM file only now. 
But with that I'm done, thanks for following up on a multi-stage review from my side

fdelzanno committed:
More style changes to avoid error. Fix handling of ldmx(log) and ldmx(trace) cases in DQM

fdelzanno committed:
Merge branch 'iss1491-a-reduced-tracking' of github.com:LDMX-Software/ldmx-sw into iss1491-a-reduced-tracking

github-actions[bot] committed:
Apply clang-format

tvami commented:
Tests are running in https://github.com/LDMX-Software/ldmx-sw/actions/runs/13546924629/job/37860769076?pr=1635

fdelzanno committed:
Fixed handling of dqm sequence

fdelzanno committed:
The real fix to the tracking dqm handling...

tvami commented on Tracking/python/reducedTracking.py:
```diff
@@ -0,0 +1,104 @@
+        self.out_seed_collection = 'LinearRecoilSeedTracks'
+        self.recoil_uncertainty = [0.006, 0.12]
+        self.ecal_uncertainty = 3.87
+        self.ecal_first_layer_z_threshold = 250
```
```suggestion
        self.ecal_first_layer_z_threshold = 250.0
```

tvami commented on Tracking/python/reducedTracking.py:
```diff
@@ -0,0 +1,104 @@
+        self.input_rec_hits_collection = 'EcalRecHits'
+        self.out_track_collection = 'LinearRecoilTruthTracks'
+        self.recoil_truth_uncertainty = [0.006, 0.12]
+        self.ecal_first_layer_z_threshold = 250
```
```suggestion
        self.ecal_first_layer_z_threshold = 250.0
```

tvami reviewed (changes requested):
Move ecal_first_layer_z_threshold to be a double

tvami committed:
Move ecal_first_layer_z_threshold to be a double

tvami commented:
@fdelzanno I think the `Local to global` printout should go into `ldmx_log(trace)` since it adds 7 lines per hit. Otherwise it seem to ran fine

fdelzanno commented:
> @fdelzanno I think the `Local to global` printout should go into `ldmx_log(trace)` since it adds 7 lines per hit. Otherwise it seem to ran fine

Ah yes, that actually could be a relic from the days (over the summer) in which we were debugging the ACTS transformations. I can covert it to a `ldmx_log(trace)`, but it looks like the info it outputs may not be super relevant -- so I could also just remove those output lines in the code.

tvami commented:
> @fdelzanno I think the `Local to global` printout should go into `ldmx_log(trace)` since it adds 7 lines per hit. Otherwise it seem to ran fine

Nvm, that's not in this PR (it's the digitizer), so that's something to change in a next PR

fdelzanno commented:
The first big change is to the LinearTruthTracking code, which has been modified to better represent a "truth track" in the rLDMX. The highlights are:
- I changed the hit collection to RecoilSimHits. Initially, I used DigiRecoilSimHits as the truth hits, but these hits are already "smeared" with respect to the actual RecoilSimHits (so they are already slightly different from the true sim-level hits).
-  I removed uncertainty from the line-fitting algorithm. The sim hits provide an exact position, so it didn't make sense to assign the realistic sensor uncertainty to them. The fitting is now a basic least-squares regression on sim hits with trackID == 1, and the chi2 of the track is calculated without uncertainty.
- The new fitting algorithm also handles cases where we only have 2 or 3 sim hits with trackID == 1. Basically, as long as we have enough points to construct a line (>= 2) from the recoil electron hits, we will output a truth track. 
- Rather than constructing combinations of hits on the recoil layers and looking for the combination with trackID == 1 and truthProb == 1 (i.e. all 4 layer hits are trackID == 1), the new code just looks directly in the collection for the points with trackID == 1. This speeds up computation and removes wasteful code.

fdelzanno commented:
The second change affects how the LinearSeedFinder creates the sensor points on which to do the track fitting. 

In the ideal case, these sensor points are created by taking the first axial/stereo pair (layers 1 and 2 of the Recoil) and combining their local measurements to produce a 3D space point. We would then do the same with the second axial/stereo pair (layers 3 and 4 of the Recoil) to produce another 3D space point. We would fit a straight line to these two points, combined with an EcalRecHit for an extra degree of freedom. Since the code to combine the axial/stereo pairs is a work in progress, the temporary solution is just to take advantage of the ability of the strip sensors at each layer to accurately measure the x-coordinate:
We still average the z-position and x-position taken from the DigiRecoilSimHits collection to find the midpoint between the axial and stereo pairs, but we now manually set the y_position to 0 (since it is not actually measured).
- We still assign an uncertainty of sigma_x = 0.006 mm (6 microns) to the x_position, but now assign a large uncertainty of sigma_y = 20/sqrt(12), where 20 is the length of the strip above/below 0.

fdelzanno committed:
Improved handling of TruthTracks. LinearSeedFinder now does not have any information about the y coordinate, since this is not measured. Fitting is mainly done in x, and the y value is left as 0 with a large uncertainty. This will be fixed once the axial/stereo pair code is debugged.

fdelzanno committed:
Resolve merge conflicts

github-actions[bot] committed:
Apply clang-format

tvami commented:
Let's run the tests and then we can merge this

fdelzanno commented:
I re-ran a 5,000 event sample using a 4GeV eGun in the reduced-ldmx-v1 geometry to show the changes in tracking from the initial comment at the beginning of the PR. Here are the statistics:

There are 5000 events in the sample.
We find 5076 tracks / 4708 events with tracks + 292 empty events in which we don't find a track = 5076 tracks / 5000 events.
We find 4862 truth tracks / 4862 events with truth tracks + 138 empty events in which we don't fit a truth track = 4862 tracks / 5000 events

Out of the tracks in the sample, 4785 tracks have trackID = 1 + 291 tracks have a different trackID number = 5076 tracks

Out of the tracks with trackID = 1 in the sample:
- there are 4544 that have truthProb = 1 (i.e. all 4 hits used in tracking correspond to the recoil electron)
- 152 tracks that have truthProb = 0.75 (i.e. 3 hits / 4 used in tracking correspond to recoil electron)
- 88 tracks that have truthProb = 0.5 (i.e. 2 hits / 4 used in tracking correspond to recoil electron)
- 292 tracks with another truthProb value

We reconstruct 4951 electron tracks + 124 positron tracks + 1 pion track = 5076 tracks.

Here are the summary plots. In blue are all the reconstructed tracks. In green are the truth tracks. In red is the data only from reconstructed tracks with trackID = 1 (i.e. reconstructed recoil electron tracks).

![eGun_rLDMX_numTrack_COMPARISON_v1](https://github.com/user-attachments/assets/0f59861d-eb76-4764-a9d7-44cff8aef5ce)

![eGun_rLDMX_chi2_COMPARISON_v1](https://github.com/user-attachments/assets/de9237c0-1404-402d-b1cc-60d98f95a0ec)

The chi2 is even lower now, which makes sense since we are fitting well in x, while the y coordinate has a large uncertainty (making it contribute less to the chi2).

![eGun_rLDMX_dRecHit_COMPARISON_v1](https://github.com/user-attachments/assets/50f5a7d0-9c5e-4ef1-997a-8e056b2a6c43)

This is the distance between the track extrapolated to the first Ecal layer and the RecHit used to create the track. The threshold value is 15 mm, so there is no reconstructed track with dRecHit > 15 mm. The radius of an Ecal cell is approximately 4 mm, meaning any track with dRecHit < 4 mm is likely to be in the correct cell. The distance from RecHit shrunk with the new tracking, probably because it is now our only reference value in y -- allowing the fitting to tune a track to fit this point well. 

I also have the pulls of the reconstructed track beam (theta) and azimuthal (phi) angle (these come from the dqm, and are for unique tracks):

<img width="1035" alt="pull_phi" src="https://github.com/user-attachments/assets/39f8a859-e86c-4a7a-81c9-75c3e4a9e1fd" />

![eGun_rLDMX_phi_COMPARISON_v1](https://github.com/user-attachments/assets/2ca5fef8-9f0f-4b54-b109-2256ed3e2833)

Pull phi and the values of phi themselves are reasonable.

<img width="1023" alt="pull_theta" src="https://github.com/user-attachments/assets/c5015502-8685-460e-850f-827745c65f98" />

![eGun_rLDMX_theta_COMPARISON_v1](https://github.com/user-attachments/assets/4b15b91b-ce98-4e09-81c1-05655cebbbf6)

Without any y information we will seriously struggle to determine a meaningful theta.

For events in which we construct a truth track, I plotted the difference between the location of the trackID = 1 reconstructed track and the truth track at both the Target and Ecal face.

![eGun_rLDMX_target_locX_diff_v1](https://github.com/user-attachments/assets/8622d3c8-e58c-4c91-9c4e-adaf899a5c71)

![eGun_rLDMX_target_locY_diff_v1](https://github.com/user-attachments/assets/7b735d8c-c1a6-490d-8ab0-495cf88e3c6e)

We cannot meaningfully fit in y, so we expect the spread in y at the target to be random (based on how well the fit is able to track the EcalRecHit).

![eGun_rLDMX_ecal_locX_diff_v1](https://github.com/user-attachments/assets/c3129707-f52a-4e95-8a8b-73f14f744f88)

![eGun_rLDMX_ecal_locY_diff_v1](https://github.com/user-attachments/assets/45f458dc-55b7-4bcd-850a-f1348e5d56e4)

Again, I believe the y position difference is behaving like we expect, since the tracking is only doing anything meaningful in y.

Overall, I think that the updated tracking is working properly and conforming to our expectations for tracks constructed without a real y measurement in the Recoil layers. Here are two example event displays to show how our tracking reconstructs tracks:

![rLDMX_v1_eventDisplay_eventID_10](https://github.com/user-attachments/assets/589f7ab2-1e1b-4318-b2c1-7941f40f3733)

![rLDMX_v1_eventDisplay_eventID_1883](https://github.com/user-attachments/assets/89af33d6-0570-47b4-87c0-0ca2ebaec5ba)

tvami commented:
> Wandering in at a late point -- has it been considered to make use of the information from the TS? It doesn't have _great_ resolution in Y, but it has some...

We tried this, unfortunately it just added "noise" to the fit

fdelzanno commented:
> > Wandering in at a late point -- has it been considered to make use of the information from the TS? It doesn't have _great_ resolution in Y, but it has some...
> 
> We tried this, unfortunately it just added "noise" to the fit

Here are some example EventDisplays from when we were trying this. 

![rLDMX_v2_tsTrack_eventDisplay_eventID_4880](https://github.com/user-attachments/assets/cc3da617-504b-49ec-826a-ab74c17e348e)

![rLDMX_v2_tsTrack_eventDisplay_eventID_4841](https://github.com/user-attachments/assets/a3d22da4-a4cd-451a-bc39-4a093e2266c9)

At the time we did conclude that the TS was just adding noise, but our fitting was also generally better pre-TS since we had y information in the fit. Perhaps it could be worth doing a quick check to see how this new tracking performs with the TS reference point in y.

tvami commented:
> Perhaps it could be worth doing a quick check to see how this new tracking performs with the TS reference point in y.

Yeah I agree. But we can do it in a next PR as an improvement to what you have here.

tvami commented:
This already ran fine:
<img width="833" alt="Screenshot 2025-03-07 at 08 57 00" src="https://github.com/user-attachments/assets/fc6bf41b-25ad-4d46-a25c-743c6f32e044" />

So feel free to merge it @fdelzanno (when you squash make sure the final commit msg makes sense and doesnt have all the back and forth msgs from all the 56 commits)
